### PR TITLE
8294696 - BufferedInputStream.transferTo should drain buffer when mark set

### PR DESCRIPTION
**Pull request by mkarg**
This PR implements JDK-8294696.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294696](https://bugs.openjdk.org/browse/JDK-8294696): BufferedInputStream.transferTo should drain buffer when mark set

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10525/head:pull/10525` \
`$ git checkout pull/10525`

Update a local copy of the PR: \
`$ git checkout pull/10525` \
`$ git pull https://git.openjdk.org/jdk pull/10525/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10525`

View PR using the GUI difftool: \
`$ git pr show -t 10525`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10525.diff">https://git.openjdk.org/jdk/pull/10525.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
This change leaks a reference to the internal buffer to the wrapped class so I think we have to be cautious and make sure that it doesn't introduce any security concerns. Also you've changed the check for a mark in a subtle way - the spec is very clear that -1 is the value when there is no mark position, values less than -1 are not defined.

**Comment by mkarg** [comment]
I think the leak is actually not a problem. `ByteArrayInputStream::transferTo` does the same BTW, see https://github.com/openjdk/jdk/blob/68da02c7b536799ccca49e889c22f3e9a2691fb7/src/java.base/share/classes/java/io/ByteArrayInputStream.java#L206-L211
In case the wrapped stream would write into the leaked buffer, that "injected" data would be lost, as no replay can happen due to the `mark == -1`  safety check. So unless we remove that check (which is not planned), such "injected" data is never read.

The JavaDoc of `markpos` literally says:

**Comment by AlanBateman** [comment]
> I think the leak is actually not a problem. 

BAIS is a fixed size. This is different to BIS where it wraps an input stream that may be changing, e.g. write an input stream to a file that is growing and call transferTo several times then you'll see what I mean. We need to mark sure we look at this very closely, that is all I'm saying.

> Regarding the `-1` check: I did that to align it with _all other_ checks of `markpos` in the existing source code of BIS. I can undo that change, but following your warning, shouldn't we fix _all other_ checks (independent of this PR) from `markpos < 0` to `markpos == -1` then?

Probably yes.

**Comment by mkarg** [comment]
> > Regarding the `-1` check: I did that to align it with _all other_ checks of `markpos` in the existing source code of BIS. I can undo that change, but following your warning, shouldn't we fix _all other_ checks (independent of this PR) from `markpos < 0` to `markpos == -1` then?
> 
> Probably yes.

I propose that I open another PR for that as those code locations are unrelated to the current PR, and this cross-method change might make the review more complex.

**Comment by Sarawin9** [comment]
> This PR implements JDK-8294696.
> 
> 
> ---------
> ### Progress
> - [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
> - [x] Change must not contain extraneous whitespace
> - [x] Commit message must refer to an issue
> 
> ### Issue
>  * [JDK-8294696](https://bugs.openjdk.org/browse/JDK-8294696): BufferedInputStream.transferTo should drain buffer when mark set
> 
> 
> ### Reviewing
> <details><summary>Using <code>git</code></summary>
> 
> Checkout this PR locally: \
> `$ git fetch https://git.openjdk.org/jdk pull/10525/head:pull/10525` \
> `$ git checkout pull/10525`
> 
> Update a local copy of the PR: \
> `$ git checkout pull/10525` \
> `$ git pull https://git.openjdk.org/jdk pull/10525/head`
> 
> </details>
> <details><summary>Using Skara CLI tools</summary>
> 
> Checkout this PR locally: \
> `$ git pr checkout 10525`
> 
> View PR using the GUI difftool: \
> `$ git pr show -t 10525`
> 
> </details>
> <details><summary>Using diff file</summary>
> 
> Download this PR as a diff file: \
> <a href="https://git.openjdk.org/jdk/pull/10525.diff">https://git.openjdk.org/jdk/pull/10525.diff</a>
> 
> </details>
>

**Comment by mkarg** [comment]
@bplb Kindly requesting review. 😃

**Comment by bplb** [comment]
Isn't a test needed here which fails without but passes with the proposed change?

**Comment by mkarg** [comment]
> Isn't a test needed here which fails without but passes with the proposed change?

I do not see such a *need*, as this is solely a performance improvement, not a fixed bug or added functionality. If the majority of this team wants to have such a test, then please clearly instruct me to write one. Thanks.

**Comment by mkarg** [comment]
> > Isn't a test needed here which fails without but passes with the proposed change?
> I do not see such a _need_, as this is solely a performance improvement, not a fixed bug or added functionality. If the majority of this team wants to have such a test, then please clearly instruct me to write one. Thanks.

@AlanBateman WDYT? Is such a test mandatory to include this rather simple PR?

**Comment by AlanBateman** [comment]
> @AlanBateman WDYT? Is such a test mandatory to include this rather simple PR?

I think it means checking that test/jdk/java/io/BufferedInputStream/TransferTo.java exercises this code path, I expect it should.

It might seem a "rather simple PR" but it leaks a reference to the internal buffer to the target and we need to get comfortable with that. It only does then when there is no mark (good) so this should mostly limit the concern to sources where EOF may be encountered more than once.

**Comment by mkarg** [comment]
So how to go on?

**Comment by rkennke** [comment]
> Isn't a test needed here which fails without but passes with the proposed change?

If you are saying, that a test should verify that, e.g. the wrapped OS receives the buffer upon calling BIS.transferTo() under certain circumstances, then no, I don't think such a test is useful, because that is an implementation detail. What should be tested is if BIS.transferTo() works correctly according to its contract, under all conceivable/possible/interesting conditions. If such a test already exists and exercises the changed code and verifies that it is still correct, then it should be ok IMO (and maybe mention in the PR which tests that would be), otherwise it would be useful to write and add such tests (either to this PR or to a preceding PR).

**Comment by rkennke** [comment]
> > @AlanBateman WDYT? Is such a test mandatory to include this rather simple PR?
> 
> I think it means checking that test/jdk/java/io/BufferedInputStream/TransferTo.java exercises this code path, I expect it should.
> 
> It might seem a "rather simple PR" but it leaks a reference to the internal buffer to the target and we need to get comfortable with that. It only does then when there is no mark (good) so this should mostly limit the concern to sources where EOF may be encountered more than once.

Which kind of sources would do that, and is there a way to check it and prevent it?

The whole issue points to an insufficiency in the API: it is lacking a way to transfer buffers in a read-only fashion (like NIO could do). I don't suppose we would want to change that, though, right?

**Comment by AlanBateman** [comment]
> Which kind of sources would do that, and is there a way to check it and prevent it?
> 
> The whole issue points to an insufficiency in the API: it is lacking a way to transfer buffers in a read-only fashion (like NIO could do). I don't suppose we would want to change that, though, right?

It can arise with many sources. One example is a program writing to a file and another reading from the same file and doing the equivalent of "tail -f". So yes, it is possible for transferTo to transfer >0 bytes after it a previous innovation has returned 0.

**Comment by mkarg** [comment]
@AlanBateman To progress with this PR, what do you expect me to do? Do you want me to provide a test that proofs that this particular example will work fine?

**Comment by AlanBateman** [comment]
The test is not the issue, it's that the proposal will require a security review. Yes, copying the bytes would remove that concern.

**Comment by mkarg** [comment]
> The test is not the issue, it's that the proposal will require a security review. Yes, copying the bytes would remove that concern.

Does "security review" mean, that I shall proof the absence of the problem, is does that term mean a formal process in the OpenJDK organization (and how do I trigger it)?

**Comment by AlanBateman** [comment]
> Does "security review" mean, that I shall proof the absence of the problem, or does that term mean a formal process in the OpenJDK organization (and how do I trigger it)?

I sent a link to this PR to one of the security engineers and they share the concern. Have you done any performance testing with an implementation that makes a defensive copy?

**Comment by mkarg** [comment]
> I sent a link to this PR to one of the security engineers and they share the concern. Have you done any performance testing with an implementation that makes a defensive copy?

Thank you. So far I have not measured the actual performance drop, as I thought it would be common sense to not drop performance *at all*, unless actually *needed*. Given the fact that the buffer could be huge depending on the caller's settings, it is hard to give a single number. For small buffers (like some KB) it is obviously neglectible, but for huge buffers (like GB) it might be drastic, and might lead to OOME in some border cases. That is why I would prefer to abstain from a defensive copy unless *needed*. :-)

Roman, your are right, the fact that the Java language misses a read-only flag (like `const` in C++) is a performance showstopper. ;-)

**Comment by mkarg** [comment]
@AlanBateman Can you please clarify: Does your answer mean that I shall provide a proof that the actual code actually does not run into the security concern, or does it mean that I *must* do defensive copy?

**Comment by AlanBateman** [comment]
> Thank you. So far I have not measured the actual performance drop, as I thought it would be common sense to not drop performance _at all_, unless actually _needed_. Given the fact that the buffer could be huge depending on the caller's settings, it is hard to give a single number. For small buffers (like some KB) it is obviously neglectible, but for huge buffers (like GB) it might be drastic, and might lead to OOME in some border cases. That is why I would prefer to abstain from a defensive copy unless _needed_. :-)

I assume you can do some experiments to see if there is a useful threshold to use. If the number of buffered bytes is larger then just delegate to the super's transferTo as before. If there isn't profitable then we may have to think about closing the PR, it's okay to have tried.

**Comment by mkarg** [comment]
I think before we give up, we should consider the alternatives, as the performance benefit of this PR is just too big to give up!

* Alternative A: I check the existing code to proof that we do not need any safety means if that is OK for everybody (is it?).

* Alternative B: Instead of double-buffering I drop the original buffer and use a same-size replacement each time the buffer was drained inside of `transferTo`. This is a rather cheap solution and rather perfectly prevents OOME, as I drop *first* before reallocating.

* Alternative C: Instead of double-buffering I zero-out (using `Arrays.fill()`) the original buffer after draining it. This would be a super-cheap alternative if somebody knows a hardware-accelerated array cleanup (do you?), but unfortunately the current code seems to be a simple Java-loop, actually.

I do not agree to close this PR unless we have a proof that the security problem really exists (I still can check that) *and* we didn't find a well-performing workaround.

**Comment by AlanBateman** [comment]
> * Alternative B: Instead of double-buffering I drop the original buffer and use a same-size replacement each time the buffer was drained inside of `transferTo`. This is a rather cheap solution and rather perfectly prevents OOME, as I drop _first_ before reallocating.

That is a good idea to try. It wouldn't work when BIS is extended but transferTo already checks this.

**Comment by mkarg** [comment]
> > * Alternative B: Instead of double-buffering I drop the original buffer and use a same-size replacement each time the buffer was drained inside of `transferTo`. This is a rather cheap solution and rather perfectly prevents OOME, as I drop _first_ before reallocating.
> 
> That is a good idea to try. It wouldn't work when BIS is extended but transferTo already checks this.

Correct, so I will give this one a try. Stay tuned. :-)

**Comment by mkarg** [comment]
@AlanBateman I have implemented Alternative B as proposed with one single change: I do not zero-out anything as in our particular case (count == pos and markpos == -1) it is cheaper to simply set count and pos both to 0 (fill() did the same trick already). Due to that any subsequent read() invocation will in turn fill(), which in turn writes valid bytes into the new buffer. So this solution should safe us from possible OOME *and* poisoned buffers. WDYT?

**Comment by AlanBateman** [comment]
> @AlanBateman I have implemented Alternative B as proposed with one single change: I do not zero-out anything as in our particular case (count == pos and markpos == -1) it is cheaper to simply set count and pos both to 0 (fill() did the same trick already). Due to that any subsequent read() invocation will in turn fill(), which in turn writes valid bytes into the new buffer. So this solution should safe us from possible OOME _and_ poisoned buffers. WDYT?

It helps a bit but it still leaks the bytes in 0 to pos, and count buf.length. I think we have to assume that Dr. Evil's output stream will throw an exception so the code to replace the buffer won't run. This means replaces buf before handing out the original buffer. The transferTo method transfers all bytes to EOF so you may be able to get away with just allocating a 0 or tiny buffer, it can grow if needed with subsequent reads or transfers.

**Comment by mkarg** [comment]
> It helps a bit but it still leaks the bytes in 0 to pos, and count buf.length. I think we have to assume that Dr. Evil's output stream will throw an exception so the code to replace the buffer won't run. This means replaces buf before handing out the original buffer. The transferTo method transfers all bytes to EOF so you may be able to get away with just allocating a 0 or tiny buffer, it can grow if needed with subsequent reads or transfers.

Silly me, you are certainly right! I have modified the code as you proposed, so now we should be safe, finally. :-)

**Comment by AlanBateman** [comment]
> Silly me, you are certainly right! I have modified the code as you proposed, so now we should be safe, finally. :-)

The updated version (7350a533) looks safe but it makes for a more complicated transferTo so it would be good to see updated performance results. One thing to look at is dropping the allocation of the new buffer after the writing of the buffered bytes. The transferTo method will read to EOF and should be rare to continue reading after EOF - if there is reading after EOF then this should cause the internal buffer to resize. The optimization is limited to the case where BIS is not extended so there are no subclasses that can observe that the internal buf has shrunk.

**Comment by mkarg** [comment]
**TL;DR:** Your optimization is now contained in this PR and outperforms the previous code. Nevertheless I plead for Alternative C (double buffering) instead, which is even faster but much simpler.

**Justification**

I took some time and did intensive JMH testing (with two buffer sizes and with four percental start positions within the buffer to simulate non-pre-read, slightly-pre-read, mostly-pre-read and fully-pre-read buffer situations) with three variants: The current code with early buffer resize (which you say is not necessary), the current code without early buffer resize (which you propose instead), and the so-called Alternative C (which uses double-buffering instead of buffer-replacing).

* Alternative B: Optimization outperforms unoptimized code
The results are shown below, and as you can see, early buffer resize is only a problem in *some*, but not all, cases. Nevertheless I agree that should be avoided, so I already pushed it as commit https://github.com/openjdk/jdk/pull/10525/commits/9c2f502c06083694deefe6a9d4c8ee4b77a92c68 and hope that you are fine with this latest piece of code.

* Alternative C: Outperforms Alternative B (optimized *and* unoptimized code) unless buffer is huge
Interesting to see BTW is that Alternative C (double-buffering instead of buffer-replacing) even outperforms that optimized code for the default buffer size (8K), and is only slightly slower in the case of large (1M) buffers (apparently `System.arraycopy` is faster than `Arrays.fill` unless the array is huge, which I did not expect originally). As the code is super-simple compared to the current code, and as the originally assumed risk of OOME does not exists (if it would, the `fill()` method would fail already, as it uses double-buffering, too, and is called by the `read` happening *before* `transferTo`) I would propose that we go with this code instead of the optimized Alternative B:
```java
private long implTransferTo(OutputStream out) throws IOException {
    if (getClass() == BufferedInputStream.class && markpos == -1) {
        int avail = count - pos;
        if (avail > 0) {
            byte[] buffer = new byte[avail]; // Double buffering prevents poisoning and leaking
            System.arraycopy(getBufIfOpen(), pos, buffer, 0, avail);
            out.write(buffer);
            pos = count;
        }
        return avail + getInIfOpen().transferTo(out);
    } else {
        return super.transferTo(out);
    }
}
```

Here are the numbers in full length for your reference:
```
Alternative B optimized (no early buffer resize)

Benchmark                                            (bufSize)  (posPercent)   Mode  Cnt        Score       Error  Units
BufferedInputStreamTransferToPerformance.transferTo       8192             0  thrpt   25     4289,949 ┬▒   217,452  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            10  thrpt   25     3823,443 ┬▒   314,171  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            90  thrpt   25     4122,949 ┬▒    52,131  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192           100  thrpt   25     4675,491 ┬▒    46,194  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576             0  thrpt   25     3477,497 ┬▒   111,229  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            10  thrpt   25     4767,220 ┬▒    18,274  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            90  thrpt   25    43900,923 ┬▒   799,471  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576           100  thrpt   25  1680163,414 ┬▒ 29823,617  ops/s

Alternative B original (with early buffer resize)

Benchmark                                            (bufSize)  (posPercent)   Mode  Cnt        Score       Error  Units
BufferedInputStreamTransferToPerformance.transferTo       8192             0  thrpt   25     4346,835 ┬▒   273,140  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            10  thrpt   25     3816,744 ┬▒   220,669  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            90  thrpt   25     3624,021 ┬▒   340,279  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192           100  thrpt   25     4484,060 ┬▒    95,201  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576             0  thrpt   25     3507,334 ┬▒   140,604  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            10  thrpt   25     4800,050 ┬▒    44,987  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            90  thrpt   25    44327,841 ┬▒   256,322  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576           100  thrpt   25  1699685,163 ┬▒ 44257,421  ops/s

Alternative C: Double buffering

Benchmark                                            (bufSize)  (posPercent)   Mode  Cnt        Score       Error  Units
BufferedInputStreamTransferToPerformance.transferTo       8192             0  thrpt   25     4555,523 ┬▒    83,370  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            10  thrpt   25     3972,562 ┬▒   489,380  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            90  thrpt   25     4181,716 ┬▒    49,527  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192           100  thrpt   25     4602,409 ┬▒   125,641  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576             0  thrpt   25     3373,734 ┬▒   161,989  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            10  thrpt   25     4684,583 ┬▒    87,257  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            90  thrpt   25    44082,232 ┬▒   364,210  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576           100  thrpt   25  1690855,778 ┬▒ 53885,439  ops/s
```

**Comment by AlanBateman** [comment]
> **TL;DR:** Your optimization is now contained in this PR and outperforms the previous code. Nevertheless I plead for Alternative C (double buffering) instead, which is even faster but much simpler.

You had dismissed this previously so it's useful to have results now to compare. I agree it's much simpler, it can probably use Arrays.copyOfRange too.

For the results, can you say what the underlying input stream is and what the target output stream is? Also it would be useful to get a baseline. I suspect you have that, it's just not pasted in the table above.

**Comment by mkarg** [comment]
> > **TL;DR:** Your optimization is now contained in this PR and outperforms the previous code. Nevertheless I plead for Alternative C (double buffering) instead, which is even faster but much simpler.

> You had dismissed this previously so it's useful to have results now to compare. I agree it's much simpler, it can probably use Arrays.copyOfRange too.

Sorry for that. Unless I actually tested, I was unaware how slow Alternative B actually is (in fact it was quite surprising that `Arrays.fill` apparently is implemented in pure Java while `System.arraycopy` is native). So let's go on with Alternative C + `Arrays.copyOfRange`! :-)

> For the results, can you say what the underlying input stream is and what the target output stream is? Also it would be useful to get a baseline. I suspect you have that, it's just not pasted in the table above.

The underlying input stream is `ByteArrayInputStream` wrapping 1 MB byte array, the output stream is `ByteArrayOutputStream`. So the test is not I/O-bound, hence it is clearly focusing on the actual difference Alternatives B/C. Yes, I do have a baseline. It is the original code without *any* safety against poisoning and leaking and the numbers are shown below. I have not published it before, because they are not perfectly comparable: A Windows Update happened after the above runs, and before this run, and my laptop feels to behave somewhat different since then. Sorry for that, but I had no time to repeat the full four hours test suite, and I think the numbers are good enough for our decision pro Alternative C. 
```
Benchmark                                            (bufSize)  (posPercent)   Mode  Cnt        Score       Error  Units
BufferedInputStreamTransferToPerformance.transferTo       8192             0  thrpt   25     4197,126 ┬▒   627,973  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            10  thrpt   25     2615,802 ┬▒   251,678  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192            90  thrpt   25     3140,941 ┬▒   575,787  ops/s
BufferedInputStreamTransferToPerformance.transferTo       8192           100  thrpt   25     4603,691 ┬▒    12,127  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576             0  thrpt   25     3539,737 ┬▒    59,913  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            10  thrpt   25     4790,750 ┬▒    63,090  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576            90  thrpt   25    43273,624 ┬▒  1784,369  ops/s
BufferedInputStreamTransferToPerformance.transferTo    1048576           100  thrpt   25  1693553,210 ┬▒ 21249,108  ops/s
```

**Comment by mkarg** [comment]
@AlanBateman Alternative C+ Arrays.copyOfRange() is implemented in https://github.com/openjdk/jdk/pull/10525/commits/92ed32cd28dab1ef5a03c6a8b1e13fef04a52753. So I think we finally got it? :-)

**Comment by AlanBateman** [comment]
> @AlanBateman Alternative C+ Arrays.copyOfRange() is implemented in 92ed32c. So I think we finally got it? :-)

Yes, I think so, this is a lot simpler, so I think you've got it to a good place.

**Review by bplb — APPROVED** [review]
Looks all right.

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/java/io/BufferedInputStream.java [hunk 1]
```diff
@@ -604,9 +605,15 @@ public long transferTo(OutputStream out) throws IOException {
...
 
     private long implTransferTo(OutputStream out) throws IOException {
-        if (getClass() == BufferedInputStream.class
-                && ((count - pos) <= 0) && (markpos == -1)) {
-            return getInIfOpen().transferTo(out);
+        if (getClass() == BufferedInputStream.class && markpos == -1) {
+            int avail = count - pos;
+            if (avail > 0) {
+                // Prevent poisoning and leaking of buf
+                byte[] buffer = Arrays.copyOfRange(getBufIfOpen(), pos, count);
```
Comment (vlsi) [inline src/java.base/share/classes/java/io/BufferedInputStream.java:612]: @mkarg , could you please clarify why you added `Arrays.copyOfRange` here?
It seems to be an excessive copy that doesn't help much.

`buf` is `protected` in `BufferedInputStream`, so if someone really wants to get hold of the actual buffer, they can subclass `BufferedInputStream` and expose the buffer directly.
What do you think of removing `copyOfRange`?

Comment (vlsi) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:612]: Buffer copy was not there before, and defensive copy was never present in `ByteArrayInputStream` as well: https://github.com/openjdk/jdk/blob/9a6ca233c7e91ffa2ce9451568b3be88ccd04504/src/java.base/share/classes/java/io/ByteArrayInputStream.java#L213

Comment (mkarg) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:612]: Alan asked for this, and for good reason: While we implicitly trust subclasses as the buffer is "theirs" as part of the inheritance, we do not know where target stream comes from. It could be an injected verhicle to perform (at least) the following:
* Leaking data. The target stream could access data beyond the intended region.
* Poisoning. The target stream could write into the buffer.

Comment (bplb) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:612]: > As Alan pointed out, it is a bug (actually even a security risk), so BAIS should get fixed, too.

I am going to file an issue on this.

Comment (mkarg) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:612]: > > As Alan pointed out, it is a bug (actually even a security risk), so BAIS should get fixed, too.
> 
> I am going to file an issue on this.

Thank you, I just planned to fix this today when I saw your existing PR! 👍
